api/src/main/java/org/openmrs/CohortMembership.java

```diff
@@ -99,6 +118,20 @@ public void setEndDate(Date endDate) {
 
 	@Override
 	public int compareTo(CohortMembership o) {
-		return this.getPatient().getPatientId() - o.getPatient().getPatientId();
+		if (Objects.equals(this.getPatient().getPatientId(), o.getPatient().getPatientId())
+				&& Objects.equals(this.getCohort().getCohortId(), o.getCohort().getCohortId())
+				&& this.getStartDate().equals(o.getStartDate())
+				&& ((this.getEndDate() != null && o.getEndDate() != null
+				&& this.getEndDate().equals(o.getEndDate()))
+				|| (this.getEndDate() == null && o.getEndDate() == null))) {
+			return 0;
+		} else if ((this.isMembershipActive() && !o.isMembershipActive())
+				|| (this.startDate.before(o.startDate))) {
+			return -1;
+		} else if ((!this.isMembershipActive() && o.isMembershipActive())
+				|| (this.startDate.after(o.startDate))) {
+			return 1;
+		}
+		return -1;
```

Comment:
You should split up each of the last 2 `else if` clauses above so that they only deal with active vs no active memberships, replace this last return statement  by delegating to OpenmrsUtil by calling `return OpenmrsUtil.compare(Date, Date)`